sycl/source/detail/platform_impl.cpp

```diff
@@ -26,6 +27,50 @@ namespace detail {
 
 using PlatformImplPtr = std::shared_ptr<platform_impl>;
 
+// Return true if the given device 'Dev' matches with any filter
+bool isForcedDevice(detail::device_filter_list *FilterList,
+                    const plugin &Plugin, const RT::PiDevice &Dev,
+                    unsigned int Index) {
+  backend Backend = Plugin.getBackend();
+  info::device_type Type;
+  Plugin.call<PiApiKind::piDeviceGetInfo>(
+      Dev, pi::cast<RT::PiDeviceInfo>(info::device::device_type), sizeof(Type),
+      &Type, nullptr);
+
+  // According to SYCL spec host device must be always available
+  if (Type == info::device_type::host)
+    return true;
+
+  for (const detail::device_filter &Filter : FilterList->get()) {
+    if ((Filter.Backend == Backend || Filter.Backend == backend::all) &&
+        (Filter.DeviceType == Type ||
+         Filter.DeviceType == info::device_type::all)) {
+      if (!Filter.HasDeviceNum ||
+          static_cast<unsigned int>(Filter.DeviceNum) == Index)
+        return true;
+    }
+  }
+  return false;
+}
+
+static void applyDeviceFilter(vector_class<RT::PiDevice> &PiDevices,
+                              const plugin &Plugin) {
+  detail::device_filter_list *FilterList =
+      detail::SYCLConfig<detail::SYCL_DEVICE_FILTER>::get();
+  if (!FilterList)
+    return;
+
+  unsigned int InsertIDx = 0;
+  for (unsigned int i = 0; i < PiDevices.size(); i++) {
```

Comment:
sycl-ls uses platform::get_devices() to get the list of devices and print them in the order of indices of the vector PiDevices.
Please note that sycl-ls tool should be run without setting SYCL_DEVICE_FILTER.
That means sycl-ls will list all possible devices in the system regardless of the env var SYCL_DEVICE_FILTER.
Now, when this function is running with SYCL_DEVICE_FILTER set, we already disabled loading unrelated plugins.
For example, sycl-ls will list three GPU devices (OCL, CUDA, L0), one CPU, one ACC, and one HOST in this order.
And the user set SYCL_DEVICE_FILTER=level_zero:3 because sycl-ls listed L0 GPU as the index number 3. 
SYCL RT will not even load open-cl and cuda plugins, so only L0 GPU device and one HOST are included in "PiDevices".
There is no index '3' in 'PiDevices' vector.
How do you handle this case?

Reply:
The warning was added to sycl-ls to as user unset the variable to get proper indexes.

Reply:
The devices are enumerated within platform. See example for 8 FPGA emulator devices and all other below:
```
$ CL_CONFIG_CPU_EMULATE_DEVICES=8 bin/sycl-ls  --verbose
Platforms: 5
Platform [#1]:
    Version  : OpenCL 1.2 Intel(R) FPGA SDK for OpenCL(TM), Version 20.3
    Name     : Intel(R) FPGA Emulation Platform for OpenCL(TM)
    Vendor   : Intel(R) Corporation
    Devices  : 8
        Device [#1]:
        Type       : ACC
        Version    : 1.2
        Name       : Intel(R) FPGA Emulation Device
        Vendor     : Intel(R) Corporation
        Driver     : 2020.11.11.0.04_160000
        Device [#2]:
        Type       : ACC
        Version    : 1.2
        Name       : Intel(R) FPGA Emulation Device
        Vendor     : Intel(R) Corporation
        Driver     : 2020.11.11.0.04_160000
        Device [#3]:
        Type       : ACC
        Version    : 1.2
        Name       : Intel(R) FPGA Emulation Device
        Vendor     : Intel(R) Corporation
        Driver     : 2020.11.11.0.04_160000
        Device [#4]:
        Type       : ACC
        Version    : 1.2
        Name       : Intel(R) FPGA Emulation Device
        Vendor     : Intel(R) Corporation
        Driver     : 2020.11.11.0.04_160000
        Device [#5]:
        Type       : ACC
        Version    : 1.2
        Name       : Intel(R) FPGA Emulation Device
        Vendor     : Intel(R) Corporation
        Driver     : 2020.11.11.0.04_160000
        Device [#6]:
        Type       : ACC
        Version    : 1.2
        Name       : Intel(R) FPGA Emulation Device
        Vendor     : Intel(R) Corporation
        Driver     : 2020.11.11.0.04_160000
        Device [#7]:
        Type       : ACC
        Version    : 1.2
        Name       : Intel(R) FPGA Emulation Device
        Vendor     : Intel(R) Corporation
        Driver     : 2020.11.11.0.04_160000
        Device [#8]:
        Type       : ACC
        Version    : 1.2
        Name       : Intel(R) FPGA Emulation Device
        Vendor     : Intel(R) Corporation
        Driver     : 2020.11.11.0.04_160000
Platform [#2]:
    Version  : OpenCL 2.1 LINUX
    Name     : Intel(R) OpenCL
    Vendor   : Intel(R) Corporation
    Devices  : 1
        Device [#1]:
        Type       : CPU
        Version    : 2.1
        Name       : Intel(R) Core(TM) i7-8700 CPU @ 3.20GHz
        Vendor     : Intel(R) Corporation
        Driver     : 2020.11.11.0.04_160000
Platform [#3]:
    Version  : OpenCL 3.0
    Name     : Intel(R) OpenCL HD Graphics
    Vendor   : Intel(R) Corporation
    Devices  : 1
        Device [#1]:
        Type       : GPU
        Version    : 3.0
        Name       : Intel(R) Graphics Gen9 [0x3e92]
        Vendor     : Intel(R) Corporation
        Driver     : 20.42.18209
Platform [#4]:
    Version  : 1.0
    Name     : Intel(R) Level-Zero
    Vendor   : Intel(R) Corporation
    Devices  : 1
        Device [#1]:
        Type       : GPU
        Version    : 1.0
        Name       : Intel(R) Graphics Gen9
        Vendor     : Intel(R) Corporation
        Driver     : 1.0.18209
Platform [#5]:
    Version  : 1.2
    Name     : SYCL host platform
    Vendor   :
    Devices  : 1
        Device [#1]:
        Type       : HOST
        Version    : 1.2
        Name       : SYCL host device
        Vendor     :
        Driver     : 1.2
default_selector()      : GPU : Intel(R) Level-Zero 1.0 [1.0.18209]
host_selector()         : HOST: SYCL host platform 1.2 [1.2]
accelerator_selector()  : ACC : Intel(R) FPGA Emulation Platform for OpenCL(TM) 1.2 [2020.11.11.0.04_160000]
cpu_selector()          : CPU : Intel(R) OpenCL 2.1 [2020.11.11.0.04_160000]
gpu_selector()          : GPU : Intel(R) Level-Zero 1.0 [1.0.18209]
custom_selector(gpu)    : GPU : Intel(R) Level-Zero 1.0 [1.0.18209]
custom_selector(cpu)    : CPU : Intel(R) OpenCL 2.1 [2020.11.11.0.04_160000]
```

When SYCL_DEVICE_FILTER is set sycl-ls reports only devices which match filter and enumerate them sequentially. The warning is added to notify user:
```
$ CL_CONFIG_CPU_EMULATE_DEVICES=8 SYCL_DEVICE_FILTER=1,4,5 bin/sycl-ls  --verbose
Warning: SYCL_DEVICE_FILTER environment variable is set. Unset it to get full list of devices and correct device indexes.
Platforms: 2
Platform [#1]:
    Version  : OpenCL 1.2 Intel(R) FPGA SDK for OpenCL(TM), Version 20.3
    Name     : Intel(R) FPGA Emulation Platform for OpenCL(TM)
    Vendor   : Intel(R) Corporation
    Devices  : 3
        Device [#1]:
        Type       : ACC
        Version    : 1.2
        Name       : Intel(R) FPGA Emulation Device
        Vendor     : Intel(R) Corporation
        Driver     : 2020.11.11.0.04_160000
        Device [#2]:
        Type       : ACC
        Version    : 1.2
        Name       : Intel(R) FPGA Emulation Device
        Vendor     : Intel(R) Corporation
        Driver     : 2020.11.11.0.04_160000
        Device [#3]:
        Type       : ACC
        Version    : 1.2
        Name       : Intel(R) FPGA Emulation Device
        Vendor     : Intel(R) Corporation
        Driver     : 2020.11.11.0.04_160000
Platform [#2]:
    Version  : 1.2
    Name     : SYCL host platform
    Vendor   :
    Devices  : 1
        Device [#1]:
        Type       : HOST
        Version    : 1.2
        Name       : SYCL host device
        Vendor     :
        Driver     : 1.2
default_selector()      : ACC : Intel(R) FPGA Emulation Platform for OpenCL(TM) 1.2 [2020.11.11.0.04_160000]
host_selector()         : HOST: SYCL host platform 1.2 [1.2]
accelerator_selector()  : ACC : Intel(R) FPGA Emulation Platform for OpenCL(TM) 1.2 [2020.11.11.0.04_160000]
cpu_selector()          : No device of requested type available. -1 (CL_DEVI...
gpu_selector()          : No device of requested type available. -1 (CL_DEVI...
custom_selector(gpu)    : No device of requested type available. -1 (CL_DEVI...
custom_selector(cpu)    : No device of requested type available. -1 (CL_DEVI...
custom_selector(acc)    : ACC : Intel(R) FPGA Emulation Platform for OpenCL(TM) 1.2 [2020.11.11.0.04_160000]
```
Also sycl-ls enumerates devices staring from 1 and SYCL_DEVICE_FILTER uses zero-based enumeration. but this should be addressed separately.

Reply:
Sorry, I still did not make myself clear. 
Let me try with an example.
Users should not add --verbose to get the desired device index.
Rather, we should ask the user to run without --verbose.
This way, they can get only the devices listed across all platforms. Here is expected output of sycl-ls with index of the device given at the beginning of each line.

> 1. ACC : Intel(R) FPGA Emulation Platform for OpenCL(TM) 1.2 [2020.11.11.0.04_160000]
2. GPU : Intel(R) OpenCL HD Graphics 3.0 [20.44.18297]
3. CPU : Intel(R) OpenCL 2.1 [2020.11.11.0.04_160000]
4. GPU : Intel(R) Level-Zero 1.0 [1.0.18297]
5. HOST: SYCL host platform 1.2 [1.2]

Now, let's assume user set SYCL_DEVICE_FILTER=level_zero:4 to set L0 GPU device.

There is no index "4" in PiDevices vector when the user program runs under SYCL_DEVICE_FILTER.
So, there is no device you can select that matches the SYCL_DEVICE_FILTER.

Reply:
By the way, Our documentation will be updated to reflect the change in sycl-ls.
The device number starts with 1, but not 0.

Reply:
The device ID is counted within platform. No cross-platform enumeration is supported. this is how it worked before my change. If this behavior need to be changed please create feature request and make sure that all agreed on the change.

Reply:
Please look at my comments below about the device number.